feat(statusline): query elapsed time and row count

CarbonChauvinist opened the pull request:
This enhances the statusline component to provide more detailed feedback on query execution.

The previous `lualine_component` is renamed to `statusline_components` and now returns a table of data instead of a simple string. This makes it more flexible for integration with other statusline plugins (e.g., heirline).

The new data includes:
- A live-updating elapsed time during query execution.
- The total elapsed time of the last completed query.
- The number of rows affected by the last query.

This is implemented by adding a `vim.loop` timer to the `query_manager` and parsing "rows affected" from LSP messages.

Includes new E2E test to ensure the timer and row count are captured correctly.

Fixes #66

CarbonChauvinist commented:
@Kurren123 Finally had time to compile these changes into a PR—would appreciate your review whenever you can..

I've been testing this locally across multiple machines for weeks and am happy with the results (it's meant to replicate the status line one gets in VSCode + mssql). That said, I recognize some changes may diverge from this project's direction. Please know I'm happy to make any modifications to better align with your goals.

The most significant change is the lualine_component refactor to support elapsed time display. Since this feature is statusline-dependent and I use Heirline locally, I restructured it to return a table of separated fields rather than a preformatted string.

This might be a breaking change for users who need to adjust their statusline configs. Below is my Heirline implementation for reference (I'll share my LuaLine config later from my Windows machine):

<details>
<summary> heirline component </summary>

```
local sl_mssql = {
  init = function(self)
    local ok, result = pcall(require, "mssql")
    if ok then
      self.components = result.statusline_components
    else
      self.components = nil
    end
  end,

  condition = function()
    return vim.b.query_manager ~= nil
  end,

  provider = function(self)
    local status_string = ""
    if not self.components or not self.components[1] then
      return ""
    end

    local state = self.components[1]()
    local format_elapsed_time = function(elapsed_time)
      return vim.fn.split(elapsed_time, "\\.")[1]
    end

    if not state or state.status == "disconnected" then
      return " Connect to MSSQL"
    end

    status_string = (state.server and " " or "")
      .. (state.server or "")
      .. (state.database and "  " or " ")
      .. (state.database or "")

    if state.status == "executing a query" then
      return string.format("%s  Executing... %s", status_string, format_elapsed_time(state.elapsed_time) or "")
    elseif state.elapsed_time then
      return string.format(
        "%s  %s  %s  %s",
        state.rows_affected or "",
        state.elapsed_time,
        status_string,
        state.caching or ""
      )
    end

    return status_string
  end,
  hl = "SignColumn",
}
```
</details>

It seems you prefer a simpler setup for potential users of this plugin, so very interested in what direction you'd like to see the statusline component taken. I'll wait for your feedback and we can take it from there.

CarbonChauvinist commented:
@Kurren123 The good news is that this is the same way that VSCode handles it - the total elapsed time is the total time of the entire script not just the last batch (see attached for screen grab):


https://github.com/user-attachments/assets/05e9498d-b65d-468b-8c57-8d20cf047463

Kurren123 commented:
Apologies, life has been busy. Will aim to take a look tonight

Kurren123 commented:
Thanks for contributing.

This is how it seems to me:

There is the "Level 1" API `lualine_component` aimed to be an easy way to satisfy most statusline users which gives them a "good enough" experience. At this moment, lualine is the most popular hence the level 1 targets lualine but of course you can still use other status lines by calling `lualine_component`.

Then there is a "Level 3" API which is harder to use, which is aimed at anyone else (lualine or otherwise) that wants to customise beyond level 1. You need to get the `query_manager` and extract any info directly from that. 

I don't think we should get rid of the Level 1 API, as it's a nice convenience for many users. So the question is do we insert a "Level 2" API (aka your `statusline_components`) which re-exposes and formats some of the `query_manager` stuff in a more easy to use way, for those in-between people who want more than what level 1 gives them but will still be happy with some sane defaults. Is it worth the extra code for those in-between users? 

I think given the number of users right now, lets leave it as these two levels of customisation and if this project grows in users we can consider an in between level at that point. In the mean time we can add a section in the docs for anyone that wants customisation beyond `lualine_component`.

**Additions to make**
That means we should add to the `query_manager` a function `last_execution` which returns a table of `row_count` and `elapsed time`. It can be updated every second while executing and you can call `vim.cmd("redrawstatus")` whenever `last_execution` set, similarly to how `new_state()` is done:

```lua
local function new_state()
	local state = states.Disconnected

	return {
		get_state = function()
			return state
		end,
		set_state = function(s)
			state = s
			vim.cmd("redrawstatus")
		end,
	}
end
```

I'm not sure if there is a need to see the last execution and also the current execution at the same time, so we can just hold the most recent in `last_excution`. This handles level 3. For level 1 you can then amend `lualine_component` to display these. 

I appreciate that for now some of the stuff in `lualine_component` will need to be duplicated in your heirline config (eg the time formatting).

Thank you again for the time you have put into this.

CarbonChauvinist committed:
feat(lualine): add live query execution stats and timer to status line

Adds live query execution time and post-execution statistics (final time and
rows affected) to the lualine status component.

- Implements a precise live timer using vim.loop.new_timer that shows elapsed
  seconds (e.g., "00:02") during query execution.
- Parses `query/complete` and `query/message` notifications within the query
  manager to capture all execution stats.
- Upon completion, updates the status with final precise time (e.g.
  "00:02.345") parsed from 'query/complete' notification.
- Listens to `query/message` for DML row counts ("N rows affected"), as they
  are not available in the `query/complete` `resultSetSummaries` used by `SELECT`.
- Adds new icon configuration options for the lualine component.

Fixes #66
